_source/_posts/2022-01-13-mobile-sso.md

```diff
@@ -0,0 +1,161 @@
+
+{% img blog/mobile-sso/ASWebAuthenticationSession-prompt.jpeg alt:"permission prompt about information sharing" width:"600" %}{: .center-image }
+
+Apple does not know whether an `ASWebAuthenticationSession` is invoked  for sign in, sign out, or general web browsing. This is why the prompt text is generic. It only states that your app is trying to *Sign In*, regardless of the actual use case, which results in the ambiguity described earlier in  **Problem 2**. 
+
+### iOS 13 changes (2019)
+
+In 2019, Apple introduced `prefersEphemeralWebBrowserSession` as an option for `ASWebAuthenticationSession`. If this option is set to true, `ASWebAuthenticationSession` does not show the prompt above, and as a consequence, it does not share cookies with Safari. This gives developers a choice. Either they gain  a better user experience without no confusing prompt and no SSO, or they get  SSO, along with the annoying prompt. 
+## `SFAuthenticationSession` or `ASWebAuthenticationSession` behavior
+The various browser options offer differing levels of cookie sharing in order to limit websites'  ability to track a user. 
+
+The cookie-sharing behavior is not well documented in Apple's documentation. The  [`SFSafariViewController` doc](https://developer.apple.com/documentation/safariservices/sfsafariviewcontroller) mentions that, "In iOS 9 and 10, it shares cookies and other website data with Safari ... If you would like to share data between your app and Safari in iOS 11 and later, ... use `ASWebAuthenticationSession` instead."
+
+A few third-party websites extend Apple's documentation to highlight that persistent cookies are shared between an embedded browser and Safari. Unfortunately, it is rarely known that session cookies can also be shared between different embedded browsers. If your application just needs SSO between your mobile apps, you do not have to use a persistent cookie; a session cookie is sufficient. 
+
+The following table summarizes the complete sharing behavior in iOS 11 or later. I've omitted `SFAuthenticationSession` for brevity because it is deprecated, but it behaves the same as `ASWebAuthenticationSession`. I've also omitted the `prefersEphemeralWebBrowserSession` option for `ASWebAuthenticationSession` because when it is set, cookies are not shared at all. 
+ 
+<div markdown="1" class="scrollable">
+
+|   | APP1 + SFSafariViewController | App1 + ASWebAuthenticationSession | App2 + SFSafariViewController | App2 + ASWebAuthenticationSession | Safari | Other browsers (e.g., Chrome) |
+|-----|-----|-----|-----|-----|-----|-----|
+| Session cookie    |   | Yes |   |  Yes |     |   |
+| Persistent cookie |   | Yes |   |  Yes | Yes |   | 
+
+</div>
```

Comment:
@mraible  can you take a quick look at this, it's similar to what you did in #994

@drhuanliu it looks like we can shrink this table down a bit by removing the duplicated App 1 and App 2 (as those values are the same, maybe something like:

|   |SFSafariViewController | ASWebAuthenticationSession | Safari | Other browsers |
|-----|-----|-----|-----|-----|
| Session cookie    | ❌ | ✅ |  ❌   | ❌ |
| Persistent cookie | ❌ | ✅ | ✅ | ❌ | 

This may also help reduce the need for the scrolling (unless browsing on mobile)

Reply:
Cannot remove app1 vs. app2, it is showing cookie can be shared between two different apps. But I like the red/green checkmarks. Maybe I can define short acronyms for SFSafariViewController and ASWebAuthenticationSession so that the heading does not go too wide.

Reply:
Maybe tweak the column title to be something like "Multiple ASWebAuthenticationSession" ?

Reply:
"Multiple ASWebAuthenticationSession" is confusing, as it may mean multiple from the same app. Trying a shorthanded version....